<!-- Copyright 2026 bitserv-ai -->
diff --git a/rocm-libraries/projects/rocblas/library/CMakeLists.txt b/rocm-libraries/projects/rocblas/library/CMakeLists.txt
index 631626fce54..de4c0101683 100644
--- a/rocm-libraries/projects/rocblas/library/CMakeLists.txt
+++ b/rocm-libraries/projects/rocblas/library/CMakeLists.txt
@@ -83,7 +83,11 @@ function( rocblas_library_settings lib_target_ )
   else()
     target_link_libraries( ${lib_target_} PRIVATE hip::device -lstdc++fs --rtlib=compiler-rt --unwindlib=libgcc)
 
-    if(BUILD_SHARED_LIBS)
+  if(NOT ROCTX)
+    add_compile_definitions("DISABLE_ROCTX")
+  endif()
+
+    if(BUILD_SHARED_LIBS AND ROCTX)
       set(ROCTX_PATH "" CACHE STRING "Path to the roctx library (directory containing libroctx64.so)")
 
       find_path(ROCTRACER_INCLUDE_DIR
