<!-- Copyright 2026 bitserv-ai -->
diff --git a/tools/completion/completion.cpp b/tools/completion/completion.cpp
index 6747558fc..ed937c9d1 100644
--- a/tools/completion/completion.cpp
+++ b/tools/completion/completion.cpp
@@ -607,7 +607,7 @@ int llama_completion(int argc, char ** argv) {
                 // - take half of the last (n_ctx - n_keep) tokens and recompute the logits in batches
 
                 if (n_past + (int) embd.size() >= n_ctx) {
-                    if (!params.ctx_shift){
+                    if (!params.ctx_shift || !llama_memory_can_shift(mem)){
                         LOG_WRN("\n\n%s: context full and context shift is disabled => stopping\n", __func__);
                         break;
                     }
@@ -637,6 +637,11 @@ int llama_completion(int argc, char ** argv) {
                 }
             } else {
                 // context extension via Self-Extend
+                if (n_past >= ga_i + ga_w && !llama_memory_can_shift(mem)){
+                    LOG_WRN("\n\n%s: context full and context shift is disabled => stopping\n", __func__);
+                    break;
+                }
+
                 while (n_past >= ga_i + ga_w) {
                     const int ib = (ga_n*ga_i)/ga_w;
                     const int bd = (ga_w/ga_n)*(ga_n - 1);
